src/core.js

```diff
@@ -12,11 +12,31 @@ jQuery.fn.init = function( selector ) {
 		migrateWarn( "jQuery( '#' ) is not a valid selector" );
 		args[ 0 ] = selector = [];
 
-	} else if ( rattrHash.test( selector ) ) {
+	// Can't use .test here because rattrHash is //g and in global scope
+	} else if ( rattrHash.exec( selector ) ) {
 
 		// The nonstandard and undocumented unquoted-hash was removed in jQuery 1.12.0
-		// Note that this doesn't actually fix the selector due to potential false positives
-		migrateWarn( "Attribute selectors with '#' must be quoted: '" + selector + "'" );
+		// First see if qS thinks it's a valid selector, if so avoid a false positive
+		try {
+			document.querySelector( selector );
+		} catch ( err1 ) {
+
+			// Didn't *look* valid to qSA, warn and try quoting what we think is the value
+			migrateWarn( "Attribute selectors with '#' must be quoted: `" + selector + "`" );
+
+			selector = selector.replace( rattrHash, function( _, attr, op, value ) {
+				return "[" + attr + op + "\"" + value + "\"]";
+			} );
+
+			// If the regexp *may* have created an invalid selector, don't update it
+			// Note that there may be false alarms if selector uses jQuery extensions
+			try {
+				document.querySelector( selector );
+				args[ 0 ] = selector;
```

Comment:
I like this "do no harm" approach, and the distinct warnings for fixable vs. unfixable selectors. But maybe we should only output one message for each bad selector, by moving the above warning to the end of this `try` block.